### ISSUE TITLE
Unable to create IAM role bindings for roleset using JSON formatting

### ISSUE THREAD

**Issue by technologik** [hashicorp/vault-plugin-secrets-gcp#27]
This issue describes inconsistencies between the `bindings` property of creating a new roleset for the GCP Secrets backend when using HCL, versus when trying to use JSON. Take the following, confirmed working, HCL config for a roleset:
```
vault write gcp/roleset/test-role \
    project="${GOOGLE_CLOUD_PROJECT}" \
    secret_type="service_account_key"  \
    bindings=@binding.hcl
```

Where `binding.hcl` is:

```
resource "//cloudresourcemanager.googleapis.com/projects/"${GOOGLE_CLOUD_PROJECT}" {
  roles = ["roles/viewer"]
}
```

When reading the config back in JSON format we receive:

```
vault read gcp/roleset/test-role -format=json
{
  "request_id": "<ID>",
  "lease_id": "",
  "lease_duration": 0,
  "renewable": false,
  "data": {
    "bindings": {
      "//cloudresourcemanager.googleapis.com/projects/<GOOGLE_CLOUD_PROJECT>": [
        "roles/viewer"
      ]
    },
    "secret_type": "service_account_key",
    "service_account_email": "vault<id-info>@<GOOGLE_CLOUD_PROJECT>.iam.gserviceaccount.com",
    "service_account_project": "<GOOGLE_CLOUD_PROJECT>"
  },
  "warnings": null
}
```

Now we use the above json to create another roleset, call it `test-role2`:
```
vault write gcp/roleset/test-role2 @test-role2.json
Success! Data written to: gcp/roleset/test-role2
```

Where the json in test-role2.json looks like:
``` 
{
    "project": "${GOOGLE_CLOUD_PROJECT},
    "secret_type": "service_account_key",
     "data": {
      "bindings": {
        "//cloudresourcemanager.googleapis.com/projects/${GOOGLE_CLOUD_PROJECT}": {
          "roles": [
            "roles/viewer"
          ]
        }
      }
    }
}
```

What we see is that while the roleset and the binding resource does get created, the list of IAM Roles is left blank:

```
vault read gcp/roleset/test-role2 -format=json
{
  "request_id": "<ID>",
  "lease_id": "",
  "lease_duration": 0,
  "renewable": false,
  "data": {
    "bindings": {
      "//cloudresourcemanager.googleapis.com/projects/<GOOGLE_CLOUD_PROJECT>": []
    },
    "secret_type": "service_account_key",
    "service_account_email": "vault<ID-info>@<GOOGLE_CLOUD_PROJECT>.iam.gserviceaccount.com",
    "service_account_project": "<GOOGLE_CLOUD_PROJECT>"
  },
  "warnings": null
}
```

Multiple permutations of the JSON were tried: both using JSON for the entire roleset config, and also only when using JSON for the bindings value itself. 

I suspect it has to do with how paths are being parsed - both line 351 and 358 [here](https://github.com/hashicorp/vault-plugin-secrets-gcp/blob/master/plugin/path_role_set.go#L351), as well as the way it is parsing the `roles` KV, [here](https://github.com/hashicorp/vault-plugin-secrets-gcp/blob/master/plugin/util/parse_bindings.go#L93), but any RCA or confirmation would be appreciated.

**Comment by jefferai** [comment]
You need `bindings` at the top level, not under `data`.

**Comment by technologik** [comment]
Hey thanks for looking at this. I originally tried keeping bindings top level (since the data object isn't mentioned in any of the docs), but has lead to more parsing errors:

remove `data` and set `bindings` as top level, using `test-role2.json`:
```
{
    "project": "${GOOGLE_CLOUD_PROJECT}",
    "secret_type": "service_account_key",
    "bindings": {
      "//cloudresourcemanager.googleapis.com/projects/${GOOGLE_CLOUD_PROJECT}": {
        "roles": [
          "roles/viewer"
        ]
      }
    }
}
```

And writing the roleset provides this error:
```
vault write gcp/roleset/test-role2 @test-role2.json
Error writing data to gcp/roleset/test-role2: Error making API request.

URL: PUT https://vault.secure.car:8200/v1/gcp/roleset/test-role2
Code: 400. Errors:

* error converting input map[//cloudresourcemanager.googleapis.com/projects/<GOOGLE_CLOUD_PROJECT>:map[roles:[roles/viewer]]] for field "bindings": '' expected type 'string', got unconvertible type 'map[string]interface {}'
```

When focusing on just the JSON for the binding value itself, for example using `binding.json`:
```
{
        "resource": {
        	"//cloudresourcemanager.googleapis.com/projects/<GOOGLE_CLOUD_PROJECT>": {
          		"roles": [
            		"roles/viewer"
          		]
        	}
        }
}
```

We get errors using an invalid KV for `roles`, which I mention in the original Issue request.
```
vault write gcp/roleset/test-role2 project="${GOOGLE_CLOUD_PROJECT}" secret_type="service_account_key" bindings=@binding.json

Error writing data to gcp/roleset/test-role2: Error making API request.

URL: PUT https://<VAULT_ADDR>/v1/gcp/roleset/test-role2
Code: 400. Errors:

* unable to parse bindings: unable to parse bindings: 1 error occurred:
	* invalid key '"roles"' in resource '//cloudresourcemanager.googleapis.com/projects/<GOOGLE_CLOUD_PROJECT>' (line 2)
```

Perhaps you could provide an example of what a proper JSON syntax might look like? I haven't seen any through my research. Thanks.

**Comment by technologik** [comment]
Hey thanks for looking at this. I originally tried keeping bindings top level (since the data object isn't mentioned in any of the docs), but has lead to more parsing errors:

remove `data` and set `bindings` as top level, using `test-role2.json`:

```
{
    "project": "${GOOGLE_CLOUD_PROJECT}",
    "secret_type": "service_account_key",
    "bindings": {
      "//cloudresourcemanager.googleapis.com/projects/${GOOGLE_CLOUD_PROJECT}": {
        "roles": [
          "roles/viewer"
        ]
      }
    }
}
```

And writing the roleset provides this error:

```
vault write gcp/roleset/test-role2 @test-role2.json
Error writing data to gcp/roleset/test-role2: Error making API request.

URL: PUT https://<VAULT_ADDR>:8200/v1/gcp/roleset/test-role2
Code: 400. Errors:

* error converting input map[//cloudresourcemanager.googleapis.com/projects/<GOOGLE_CLOUD_PROJECT>:map[roles:[roles/viewer]]] for field "bindings": '' expected type 'string', got unconvertible type 'map[string]interface {}'
```

When focusing on just the JSON for the binding value itself, for example using binding.json:
```
{
        "resource": {
        	"//cloudresourcemanager.googleapis.com/projects/<GOOGLE_CLOUD_PROJECT>": {
          		"roles": [
            		"roles/viewer"
          		]
        	}
        }
}
```

I get errors using an invalid KV for roles, which I mention in the original Issue request.
```
vault write gcp/roleset/test-role2 project="${GOOGLE_CLOUD_PROJECT}" secret_type="service_account_key" bindings=@binding.json

Error writing data to gcp/roleset/test-role2: Error making API request.

URL: PUT https://<VAULT_ADDR>/v1/gcp/roleset/test-role2
Code: 400. Errors:

* unable to parse bindings: unable to parse bindings: 1 error occurred:
	* invalid key '"roles"' in resource '//cloudresourcemanager.googleapis.com/projects/<GOOGLE_CLOUD_PROJECT>' (line 2)
```

Perhaps you could provide an example of what a proper JSON syntax might look like? I haven't seen any through my research. Thanks.

**Comment by technologik** [comment]
Following up on this. We confirmed that [this line](https://github.com/hashicorp/vault-plugin-secrets-gcp/blob/master/plugin/util/parse_bindings.go#L93) in the parse_bindings file is causing the error you see in my previous comment here:

```
unable to parse bindings: unable to parse bindings: 1 error occurred:
	* invalid key '"roles"' in resource '//cloudresourcemanager.googleapis.com/projects/<GOOGLE_CLOUD_PROJECT>' (line 2)
```

Specifically, it's checking for `"roles"` , and not `roles`. We tested this by included a line just above in the parse_bindings file to strip the quotes:
`key = strings.Replace(key, "\"", "", -1)`

And it was no longer throwing an error with the proper bindings file seen in `test-role2.json` from the comment above.
